crates/sncast/src/lib.rs

```diff
@@ -300,78 +358,86 @@ pub async fn wait_for_tx(
                 let remaining_time = i * u16::from(retry_interval);
                 println!("Waiting for transaction to be accepted ({i} retries / {remaining_time}s left until timeout)");
             }
-            Err(err) => return Err(err.into()),
+            Err(err) => return Err(WaitForTransactionError::Other(err.into())),
         };
 
         sleep(Duration::from_secs(retry_interval.into()));
     }
 
-    Err(anyhow!(
+    Err(WaitForTransactionError::Other(anyhow!(
```

Comment:
If the transaction is rejected it should have been caught earlier https://github.com/foundry-rs/starknet-foundry/pull/1710/files#diff-746993eb163f47a361248767cf9e32a0e39151785fa5f5ff2cdef29ba9e50a72R342

`not received` - what could be the reasons the transaction hash not received?

`sncast timed out` - hmm we can add a separate value to enum for this one, the user may want to handle it

Reply:
> If the transaction is rejected it should have been caught earlier

well yes, but actually no. It depends on timeout set by the user, probably

> what could be the reasons the transaction hash not received?

network errors for sure, if I remember correctly sequencer may also mark tx as invalid which may in some cases also result in not received (as tx will be rejected straight away)

>  hmm we can add a separate value to enum for this one, the user may want to handle it

Yes, this is basically a tool configuration + eventually network slowness errors? (although, not sure I would call them errors - it's just something that might happen if network is slow, or the users specified slower timeouts, because they did not want to wait indefinitely)

Reply:
https://docs.starknet.io/documentation/architecture_and_concepts/Network_Architecture/transaction-life-cycle/
Sequencer may label it with Rejected status, but mempool may label it with NOT RECEIVED

In the for loop we have `prinln` that informs the user that the transaction was not received yet. I think that in [this](https://github.com/foundry-rs/starknet-foundry/pull/1710/files#diff-746993eb163f47a361248767cf9e32a0e39151785fa5f5ff2cdef29ba9e50a72R368) line we can only inform the user that sncast timed out, since the user will already know about the current transaction status (either from println, or from the error returned from the for loop)

Reply:
> In the for loop we have prinln that informs the user that the transaction was not received yet.

I don't think that's true, as we cannot verify this. We inform that tx was not accepted, which is a different thing in itself. (that is unless I miss something, then please point it out to me)

Reply:
We can easily verify if sncast timed-out / tx is rejected though, so maybe this is the way to streamline the message a bit (then the "tx not received" would de-facto become our last case, if all other scenarios are met).

In other words - we know when tx is rejected, and when sncast times out so we can throw an errors there. We do not know when tx is not received, but we may assume that it will always be the case when the other two cases (+ success cases) are not true, right?

Reply:
we don't know if the transaction will be received. When sncast is timed out we know that the transaction has not been received yet. But maybe it will be received in the future

Reply:
I'm starting to think that we should separate these two https://github.com/foundry-rs/starknet-foundry/pull/1710/files#diff-746993eb163f47a361248767cf9e32a0e39151785fa5f5ff2cdef29ba9e50a72R356-R357

As if the transaction has been received, it may become rejected, but not `not received` https://docs.starknet.io/documentation/architecture_and_concepts/Network_Architecture/transaction-life-cycle/

Reply:
> we don't know if the transaction will be received. When sncast is timed out we know that the transaction has not been received yet. But maybe it will be received in the future

the problem with this is this is pretty non-deterministic. Maybe it will be received, but maybe not. It's really just an uneducated guess. Normally mempool/sequencer validates tx within seconds, but what if there are transient issues?

> I'm starting to think that we should separate these two https://github.com/foundry-rs/starknet-foundry/pull/1710/files#diff-746993eb163f47a361248767cf9e32a0e39151785fa5f5ff2cdef29ba9e50a72R356-R357

sadly the link just redirects to the top of this review, and not to specified lines :(

Reply:
```
Ok(starknet::core::types::TransactionStatus::Received)
            | Err(StarknetError(TransactionHashNotFound)) => {
```
This is the line I was talking about

> I'm starting to think that we should separate these two
> As if the transaction has been received, it may become rejected, but not not received

Reply:
So, I imagine the flow for the transaction that has not been received yet, the user will see the message "Waiting for transaction to be accepted..." and then sncast timed out. 

So instead of
```
Failed to get transaction with hash = {tx_hash:#x}; Transaction rejected, not received or sncast timed out
```
we can have 
```
Err(TimedOut)
```

Reply:
changed https://github.com/foundry-rs/starknet-foundry/pull/1710/commits/7d272ac511692fc9a01e54592dac9d8a3f0f6f3a

Reply:
and https://github.com/foundry-rs/starknet-foundry/pull/1710/commits/fdce64431d9f457527f5f21fb945d0c27ac19467